src/components/feedback/ToastMessages.tsx

```diff
@@ -0,0 +1,219 @@
+    </Callout>
+  );
+}
+
+type ToastMessageTransitionType = FunctionComponent<
+  ComponentProps<TransitionComponent> & {
+    transitionClasses?: ToastMessageTransitionClasses;
+  }
+>;
+
+const ToastMessageTransition: ToastMessageTransitionType = ({
+  direction,
+  onTransitionEnd,
+  children,
+  transitionClasses = {},
+}) => {
+  const isDismissed = direction === 'out';
+  const containerRef = useRef<HTMLDivElement>(null);
+  const handleAnimation = (e: AnimationEvent) => {
+    // Ignore animations happening on child elements
+    if (e.target !== containerRef.current) {
+      return;
+    }
+
+    onTransitionEnd?.(direction ?? 'in');
+  };
+  const classes = useMemo(() => {
+    const {
+      transitionIn = 'animate-fade-in',
+      transitionOut = 'animate-fade-out',
+    } = transitionClasses;
+
+    return {
+      [transitionIn]: !isDismissed,
+      [transitionOut]: isDismissed,
+    };
+  }, [isDismissed, transitionClasses]);
+
+  return (
+    <div
+      data-testid="animation-container"
+      onAnimationEnd={handleAnimation}
+      ref={containerRef}
+      className={classnames('relative w-full container', classes)}
+    >
+      {children}
+    </div>
+  );
+};
+
+export type ToastMessagesProps = {
+  messages: ToastMessage[];
+  onMessageDismiss: (id: string) => void;
+  transitionClasses?: ToastMessageTransitionClasses;
+  setTimeout_?: typeof setTimeout;
+};
+
+/**
+ * A collection of toast messages. These are rendered within an `aria-live`
+ * region for accessibility with screen readers.
+ */
+export default function ToastMessages({
+  messages,
+  onMessageDismiss,
+  transitionClasses,
+  /* istanbul ignore next - test seam */
+  setTimeout_ = setTimeout,
+}: ToastMessagesProps) {
+  // List of IDs of toast messages that have been dismissed and have an
+  // in-progress 'out' transition
+  const [dismissedMessages, setDismissedMessages] = useState<string[]>([]);
+  // Tracks not finished timeouts for auto-dismiss toast messages
+  const messageSchedules = useRef(new Map<string, number>());
+
+  const dismissMessage = useCallback(
+    (id: string) => setDismissedMessages(ids => [...ids, id]),
+    [],
+  );
+  const scheduleMessageDismiss = useCallback(
+    (id: string) => {
+      const timeout = setTimeout_(() => {
+        dismissMessage(id);
+        messageSchedules.current.delete(id);
+      }, 5000);
+      messageSchedules.current.set(id, timeout);
+    },
+    [dismissMessage, setTimeout_],
+  );
+
+  const onTransitionEnd = useCallback(
+    (direction: 'in' | 'out', message: ToastMessage) => {
+      const autoDismiss = message.autoDismiss ?? true;
+      if (direction === 'in' && autoDismiss) {
+        scheduleMessageDismiss(message.id);
+      }
+
+      if (direction === 'out') {
+        onMessageDismiss(message.id);
+        setDismissedMessages(ids => ids.filter(id => id !== message.id));
+      }
+    },
+    [scheduleMessageDismiss, onMessageDismiss],
+  );
+
+  useEffect(() => {
```

Comment:
Probably want to use a layout effect for this so that it happens immediately on unmount, rather than on the next animation frame.